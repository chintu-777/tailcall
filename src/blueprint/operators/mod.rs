mod call;
mod expr;
mod graphql;
mod grpc;
mod http;
mod modify;
mod protected;
mod jq;

pub use call::*;
pub use expr::*;
pub use jq::*;
pub use graphql::*;
pub use grpc::*;
pub use http::*;
pub use modify::*;
pub use protected::*;
