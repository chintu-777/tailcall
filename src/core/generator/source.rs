use thiserror::Error;

///
/// A list of sources from which a configuration can be created
#[derive(Clone, Copy, PartialEq, Debug, Default)]
pub enum Source {
    #[default]
    Proto,
    Graphql,
}

#[derive(Debug, Error, PartialEq)]
#[error("Unsupported config extension: {0}")]
pub struct UnsupportedFileFormat(String);

impl std::str::FromStr for Source {
    type Err = UnsupportedFileFormat;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_lowercase().as_str() {
            "proto" => Ok(Source::Proto),
            "graphql" | "gql" => Ok(Source::Graphql),
            _ => Err(UnsupportedFileFormat(s.to_string())),
        }
    }
}

#[cfg(test)]
mod tests {
    use std::str::FromStr;

    use super::*;

    #[test]
    fn test_from_str() {
        assert_eq!(Source::from_str("proto"), Ok(Source::Proto));
        assert_eq!(Source::from_str("gql"), Ok(Source::Graphql));
        assert!(Source::from_str("foo").is_err());
    }
}
