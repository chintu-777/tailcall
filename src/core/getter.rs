use crate::core::config::{Enum, Type};

pub trait GetName {
    fn get_name(&self) -> &str;
}

impl GetName for Type {
    fn get_name(&self) -> &str {
        &self.name
    }
}

impl GetName for Enum {
    fn get_name(&self) -> &str {
        &self.name
    }
}

pub trait Getter<T> {
    // TODO rename
    fn get(&self, key: &str) -> Option<&T>;
    fn get_mut(&mut self, key: &str) -> Option<&mut T>;
    fn insert(self, value: T) -> Self;
    fn remove(self, key: &str) -> Self
    where
        Self: Sized;
}

impl<T: GetName> Getter<T> for Vec<T> {
    fn get(&self, key: &str) -> Option<&T> {
        self.iter().find(|t| t.get_name() == key)
    }

    fn get_mut(&mut self, key: &str) -> Option<&mut T> {
        self.iter_mut().find(|t| t.get_name() == key)
    }

    fn insert(mut self, value: T) -> Self {
        self.push(value);
        self
    }

    fn remove(mut self, key: &str) -> Self {
        self.retain(|t| t.get_name() != key);
        self
    }
}
