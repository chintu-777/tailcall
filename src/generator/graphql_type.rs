use std::fmt::Display;

use convert_case::{Case, Casing};
pub(super) static DEFAULT_SEPARATOR: &str = "__";
static PACKAGE_SEPARATOR: &str = ".";

fn normalize_name(name: &str) -> String {
    name.replace(PACKAGE_SEPARATOR, "_")
}

/// A struct to represent the name of a GraphQL type.
#[derive(Debug, Clone)]
pub struct GraphQLType<A>(A);

#[derive(Debug, Clone)]
pub struct Parsed {
    namespace: Option<Namespace>,
    name: String,
    entity: Entity,
}

#[derive(Debug, Clone)]
pub struct Unparsed {
    namespace: Option<Namespace>,
    name: String,
}

#[derive(Debug, Clone)]
struct Namespace {
    path: Vec<String>,
}

impl Namespace {
    fn parse(input: &str) -> Self {
        let path = input
            .split(PACKAGE_SEPARATOR)
            .map(String::from)
            .collect::<Vec<_>>();

        Self { path }
    }

    fn combine(mut self, other: Namespace) -> Self {
        self.path.extend(other.path);

        self
    }
}

impl Display for Namespace {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.path.join(DEFAULT_SEPARATOR).as_str())
    }
}

impl GraphQLType<Unparsed> {
    pub fn new(name: &str) -> Self {
        Self(Unparsed { namespace: None, name: name.to_string() })
    }

    pub fn append_namespace(mut self, namespace: &str) -> Self {
        if namespace.is_empty() {
            return self;
        }

        let additional = Namespace::parse(namespace);

        self.0.namespace = if let Some(namespace) = self.0.namespace {
            Some(namespace.combine(additional))
        } else {
            Some(additional)
        };

        self
    }

    fn parse(self, entity: Entity) -> GraphQLType<Parsed> {
        let unparsed = self.0;
        let name = normalize_name(&unparsed.name);
        let namespace = unparsed.namespace;

        GraphQLType(Parsed { name, namespace, entity })
    }

    pub fn as_enum(self) -> GraphQLType<Parsed> {
        self.parse(Entity::Enum)
    }

    pub fn as_enum_variant(self) -> GraphQLType<Parsed> {
        self.parse(Entity::EnumVariant)
    }

    pub fn as_object_type(self) -> GraphQLType<Parsed> {
        self.parse(Entity::ObjectType)
    }

    pub fn as_method(self) -> GraphQLType<Parsed> {
        self.parse(Entity::Method)
    }

    pub fn as_field(self) -> GraphQLType<Parsed> {
        self.parse(Entity::Field)
    }
}

impl GraphQLType<Parsed> {
    pub fn id(&self) -> String {
        if let Some(ref namespace) = self.0.namespace {
            format!("{}.{}", namespace, self.0.name)
        } else {
            self.0.name.clone()
        }
    }
}

// FIXME: make it private
/// Used to convert proto type names to GraphQL formatted names.
/// Enum to represent the type of the descriptor
#[derive(Clone, Debug)]
enum Entity {
    Enum,
    EnumVariant,
    ObjectType,
    Method,
    Field,
}

impl Display for GraphQLType<Parsed> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let parsed = &self.0;
        match parsed.entity {
            Entity::EnumVariant => f.write_str(parsed.name.as_str())?,
            Entity::Field => f.write_str(parsed.name.to_case(Case::Camel).as_str())?,
            Entity::Method => {
                if let Some(package) = &parsed.namespace {
                    f.write_str(package.to_string().as_str())?;
                    f.write_str(DEFAULT_SEPARATOR)?;
                };
                f.write_str(parsed.name.as_str())?
            }
            Entity::Enum | Entity::ObjectType => {
                if let Some(package) = &parsed.namespace {
                    f.write_str(package.to_string().as_str())?;
                    f.write_str(DEFAULT_SEPARATOR)?;
                };
                f.write_str(parsed.name.as_str())?
            }
        };
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;

    use super::*;

    type TestParams = ((Entity, Option<&'static str>, &'static str), &'static str);

    #[test]
    fn test_from_enum() {
        let input: Vec<TestParams> = vec![
            // Enums
            ((Entity::Enum, None, "foo"), "foo"),
            ((Entity::Enum, None, "a.b.c.foo"), "a_b_c_foo"),
            ((Entity::Enum, Some("a.b.c"), "foo"), "a__b__c__foo"),
            (
                (Entity::Enum, Some("a.b.c"), "d.e.f.foo"),
                "a__b__c__d_e_f_foo",
            ),
            ((Entity::Enum, Some(""), "a.b.c.foo"), "a_b_c_foo"),
            ((Entity::Enum, None, "a_b_c_foo"), "a_b_c_foo"),
        ];

        assert_type_names(input);
    }

    #[test]
    fn test_from_enum_variant() {
        let input: Vec<TestParams> = vec![
            // Enum variants
            ((Entity::EnumVariant, None, "foo"), "FOO"),
            ((Entity::EnumVariant, None, "a.b.c.foo"), "FOO"),
            ((Entity::EnumVariant, Some("a.b.c"), "foo"), "FOO"),
            ((Entity::EnumVariant, Some("a.b"), "d.e.foo"), "FOO"),
            ((Entity::EnumVariant, Some(""), "a.b.c.foo"), "FOO"),
            ((Entity::EnumVariant, None, "a_b_c_foo"), "A_B_C_FOO"),
        ];

        assert_type_names(input);
    }

    #[test]
    fn test_from_object_type() {
        let input: Vec<TestParams> = vec![
            // Object types
            ((Entity::ObjectType, None, "foo"), "FOO"),
            ((Entity::ObjectType, None, "a.b.c.foo"), "A_B_C_FOO"),
            ((Entity::ObjectType, Some("a.b.c"), "foo"), "A_B_C_FOO"),
            ((Entity::ObjectType, Some("a.b"), "d.e.foo"), "A_B_FOO"),
            ((Entity::ObjectType, Some(""), "a.b.c.foo"), "A_B_C_FOO"),
            ((Entity::ObjectType, None, "a_b_c_foo"), "A_B_C_FOO"),
            ((Entity::ObjectType, None, "foo.bar.Baz"), "FOO_BAR_BAZ"),
        ];

        assert_type_names(input);
    }

    #[test]
    fn test_from_method() {
        let input: Vec<TestParams> = vec![
            // Methods
            ((Entity::Method, None, "foo"), "foo"),
            ((Entity::Method, None, "a.b.c.foo"), "a_b_c_foo"),
            ((Entity::Method, Some("a.b.c"), "foo"), "a_b_c__foo"),
            ((Entity::Method, Some("a.b"), "d.e.foo"), "a_b__d_e_foo"),
            ((Entity::Method, Some(""), "a.b.c.foo"), "a_b_c_foo"),
            ((Entity::Method, None, "a_bC_foo"), "a_b_c_foo"),
        ];

        assert_type_names(input);
    }

    #[test]
    fn test_from_field() {
        let input: Vec<TestParams> = vec![
            // Fields
            ((Entity::Field, None, "foo"), "foo"),
            ((Entity::Field, None, "a.b.c.foo"), "foo"),
            ((Entity::Field, Some("a.b.c"), "foo"), "foo"),
            ((Entity::Field, Some("a.b"), "d.e.foo"), "foo"),
            ((Entity::Field, Some(""), "a.b.c.foo"), "foo"),
            ((Entity::Field, None, "a_bC_foo"), "a_b_c_foo"),
        ];

        assert_type_names(input);
    }

    fn assert_type_names(input: Vec<TestParams>) {
        for ((entity, namespace, name), expected) in input {
            let mut g = GraphQLType::new(name);
            if let Some(namespace) = namespace {
                g = g.append_namespace(namespace);
            }

            let actual = g.parse(entity).to_string();
            assert_eq!(actual, expected);
        }
    }
}
