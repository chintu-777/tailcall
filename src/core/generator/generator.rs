use anyhow::Result;

use crate::core::config::{Config, ConfigModule, Link, LinkType, Resolution};
use crate::core::generator::from_proto::from_proto;
use crate::core::generator::Source;
use crate::core::merge_right::MergeRight;
use crate::core::runtime::TargetRuntime;
use crate::ConfigReader;

pub struct Generator {
    config_reader: ConfigReader,
}
impl Generator {
    pub fn init(runtime: TargetRuntime) -> Self {
        let config_reader = ConfigReader::init(runtime);
        Self { config_reader }
    }
    #[allow(clippy::too_many_arguments)]
    pub async fn read_all<T: AsRef<str>>(
        &self,
        input_source: Source,
        files: &[T],
        query: &str,
        resolution: impl Fn(&str) -> Resolution,
    ) -> Result<ConfigModule> {
        let mut links = vec![];
        let mut config = Config::default();

        match input_source {
            Source::Proto => {
                let proto_metadata = self.config_reader.proto_reader.read_all(files).await?;
                for metadata in proto_metadata {
                    links.push(Link { id: None, src: metadata.path, type_of: LinkType::Protobuf });
                    config = config.merge_right(from_proto(&[metadata.descriptor_set], query));
                }
            }
            Source::Graphql => {
                let module = self.config_reader.read_all(files).await?;
                config = config.merge_right(module.config);
            }
        }

        config.links = links;
        Ok(ConfigModule::from(config).resolve_ambiguous_types(resolution))
    }
}

#[cfg(test)]
mod test {
    use std::path::PathBuf;

    use super::*;

    fn start_mock_server() -> httpmock::MockServer {
        httpmock::MockServer::start()
    }

    #[tokio::test]
    async fn test_read_all() {
        let server = start_mock_server();
        let runtime = crate::core::runtime::test::init(None);
        let test_dir = PathBuf::from(tailcall_fixtures::generator::proto::SELF);

        let news_content = runtime
            .file
            .read(test_dir.join("news.proto").to_str().unwrap())
            .await
            .unwrap();
        let greetings_a = runtime
            .file
            .read(test_dir.join("greetings_a.proto").to_str().unwrap())
            .await
            .unwrap();

        server.mock(|when, then| {
            when.method(httpmock::Method::GET).path("/news.proto");
            then.status(200)
                .header("Content-Type", "application/vnd.google.protobuf")
                .body(&news_content);
        });

        server.mock(|when, then| {
            when.method(httpmock::Method::GET)
                .path("/greetings_a.proto");
            then.status(200)
                .header("Content-Type", "application/protobuf")
                .body(&greetings_a);
        });

        let reader = Generator::init(runtime);
        let news = format!("http://localhost:{}/news.proto", server.port());
        let greetings_a = format!("http://localhost:{}/greetings_a.proto", server.port());
        let greetings_b = test_dir
            .join("greetings_b.proto")
            .to_str()
            .unwrap()
            .to_string();

        let config = reader
            .read_all(
                Source::Proto,
                &[news, greetings_a, greetings_b],
                "Query",
                |v: &str| Resolution { input: format!("IN_{}", v), output: format!("OUT_{}", v) },
            )
            .await
            .unwrap();

        assert_eq!(config.links.len(), 3);
        assert_eq!(config.types.get("Query").unwrap().fields.len(), 8);
    }
}
