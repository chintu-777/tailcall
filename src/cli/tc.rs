use std::fs;
use std::path::Path;
use std::str::FromStr;

use anyhow::Result;
use clap::Parser;
use convert_case::{Case, Casing};
use dotenvy::dotenv;
use inquire::Confirm;
use lazy_static::lazy_static;
use stripmargin::StripMargin;

use super::command::{Cli, Command};
use super::update_checker;
use crate::cli::fmt::Fmt;
use crate::cli::server::Server;
use crate::cli::{self, CLIError};
use crate::core::blueprint::Blueprint;
use crate::core::config::reader::ConfigReader;
use crate::core::config::{Resolution, ResolveOptions};
use crate::core::generator::Generator;
use crate::core::http::API_URL_PREFIX;
use crate::core::print_schema;
use crate::core::rest::{EndpointSet, Unchecked};

const FILE_NAME: &str = ".tailcallrc.graphql";
const YML_FILE_NAME: &str = ".graphqlrc.yml";
const JSON_FILE_NAME: &str = ".tailcallrc.schema.json";

lazy_static! {
    static ref TRACKER: tailcall_tracker::Tracker = tailcall_tracker::Tracker::default();
}

pub async fn run() -> Result<()> {
    let cli = Cli::parse();
    let writer = std::io::stdout().lock();
    run_inner(cli, writer).await
}

pub async fn run_inner<W: std::io::Write>(cli: Cli, mut writer: W) -> Result<()> {
    if let Ok(path) = dotenv() {
        tracing::info!("Env file: {:?} loaded", path);
    }
    update_checker::check_for_update().await;
    let runtime = cli::runtime::init(&Blueprint::default());
    let config_reader = ConfigReader::init(runtime.clone());

    // Initialize ping event every 60 seconds
    let _ = TRACKER
        .init_ping(tokio::time::Duration::from_secs(60))
        .await;

    // Dispatch the command as an event
    let _ = TRACKER
        .dispatch(cli.command.to_string().to_case(Case::Snake).as_str())
        .await;
    match cli.command {
        Command::Start { file_paths } => {
            let config_module = config_reader.read_all(&file_paths).await?;
            log_endpoint_set(&config_module.extensions.endpoint_set);
            Fmt::log_n_plus_one(false, &config_module.config);
            let server = Server::new(config_module);
            server.fork_start().await?;
            Ok(())
        }
        Command::Check { file_paths, n_plus_one_queries, schema, format } => {
            let config_module = (config_reader.read_all(&file_paths)).await?;
            log_endpoint_set(&config_module.extensions.endpoint_set);
            if let Some(format) = format {
                Fmt::display(format.encode(&config_module)?, &mut writer);
            }
            let blueprint = Blueprint::try_from(&config_module).map_err(CLIError::from);

            match blueprint {
                Ok(blueprint) => {
                    tracing::info!("Config {} ... ok", file_paths.join(", "));
                    Fmt::log_n_plus_one(n_plus_one_queries, &config_module.config);
                    // Check the endpoints' schema
                    let _ = config_module
                        .extensions
                        .endpoint_set
                        .into_checked(&blueprint, runtime)
                        .await?;
                    if schema {
                        display_schema(&blueprint, &mut writer);
                    }
                    Ok(())
                }
                Err(e) => Err(e.into()),
            }
        }
        Command::Init { folder_path } => init(&folder_path).await,
        Command::Gen {
            file_paths,
            input,
            output,
            query,
            resolve_ambiguous_input,
            resolve_ambiguous_output,
        } => {
            let generator = Generator::init(runtime);
            let (resolve_ambiguous_input, resolve_ambiguous_output) =
                validate_resolutions(resolve_ambiguous_input, resolve_ambiguous_output)?;

            let cfg = generator
                .read_all(
                    input,
                    file_paths.as_ref(),
                    query.as_str(),
                    resolver_fn(resolve_ambiguous_input, resolve_ambiguous_output),
                )
                .await?;

            let config = output.unwrap_or_default().encode(&cfg)?;
            Fmt::display(config, &mut writer);
            Ok(())
        }
    }
}

fn resolver_fn(
    resolve_ambiguous_input: Option<ResolveOptions>,
    resolve_ambiguous_output: Option<ResolveOptions>,
) -> impl Fn(&str) -> Resolution {
    move |v| {
        let mut resolution =
            Resolution { input: format!("IN_{}", v), output: format!("OUT_{}", v) };
        if let Some(ResolveOptions { prefix, suffix }) = &resolve_ambiguous_input {
            resolution.input = format!("{}{}{}", prefix, v, suffix);
        }

        if let Some(ResolveOptions { prefix, suffix }) = &resolve_ambiguous_output {
            resolution.output = format!("{}{}{}", prefix, v, suffix);
        }
        resolution
    }
}

fn validate_resolutions(
    input: Option<String>,
    output: Option<String>,
) -> Result<(Option<ResolveOptions>, Option<ResolveOptions>)> {
    let input = input
        .as_ref()
        .and_then(|v| ResolveOptions::from_str(v).ok());

    let output = output
        .as_ref()
        .and_then(|v| ResolveOptions::from_str(v).ok());

    if let Some(input) = input.as_ref() {
        if let Some(output) = output.as_ref() {
            validate_against(input, output)?;
        }
    }

    Ok((input, output))
}

// TODO move this to impl ResolveOptions once it's integrated with
// *Reader::read_all
pub fn validate_against(s: &ResolveOptions, other: &ResolveOptions) -> Result<()> {
    let lhs = format!("{}{}", s.prefix, s.suffix);
    let rhs = format!("{}{}", other.prefix, other.suffix);

    if lhs == rhs {
        return Err(anyhow::anyhow!(
            "Input and output resolutions cannot be the same"
        ));
    }
    Ok(())
}

pub async fn init(folder_path: &str) -> Result<()> {
    let folder_exists = fs::metadata(folder_path).is_ok();

    if !folder_exists {
        let confirm = Confirm::new(&format!(
            "Do you want to create the folder {}?",
            folder_path
        ))
        .with_default(false)
        .prompt()?;

        if confirm {
            fs::create_dir_all(folder_path)?;
        } else {
            return Ok(());
        };
    }

    let tailcallrc = include_str!("../../generated/.tailcallrc.graphql");
    let tailcallrc_json: &str = include_str!("../../generated/.tailcallrc.schema.json");

    let file_path = Path::new(folder_path).join(FILE_NAME);
    let json_file_path = Path::new(folder_path).join(JSON_FILE_NAME);
    let yml_file_path = Path::new(folder_path).join(YML_FILE_NAME);

    let tailcall_exists = fs::metadata(&file_path).is_ok();

    if tailcall_exists {
        // confirm overwrite
        let confirm = Confirm::new(&format!("Do you want to overwrite the file {}?", FILE_NAME))
            .with_default(false)
            .prompt()?;

        if confirm {
            fs::write(&file_path, tailcallrc.as_bytes())?;
            fs::write(&json_file_path, tailcallrc_json.as_bytes())?;
        }
    } else {
        fs::write(&file_path, tailcallrc.as_bytes())?;
        fs::write(&json_file_path, tailcallrc_json.as_bytes())?;
    }

    let yml_exists = fs::metadata(&yml_file_path).is_ok();

    if !yml_exists {
        fs::write(&yml_file_path, "")?;

        let graphqlrc = r"|schema:
         |- './.tailcallrc.graphql'
    "
        .strip_margin();

        fs::write(&yml_file_path, graphqlrc)?;
    }

    let graphqlrc = fs::read_to_string(&yml_file_path)?;

    let file_path = file_path.to_str().unwrap();

    let mut yaml: serde_yaml::Value = serde_yaml::from_str(&graphqlrc)?;

    if let Some(mapping) = yaml.as_mapping_mut() {
        let schema = mapping
            .entry("schema".into())
            .or_insert(serde_yaml::Value::Sequence(Default::default()));
        if let Some(schema) = schema.as_sequence_mut() {
            if !schema
                .iter()
                .any(|v| v == &serde_yaml::Value::from("./.tailcallrc.graphql"))
            {
                let confirm =
                    Confirm::new(&format!("Do you want to add {} to the schema?", file_path))
                        .with_default(false)
                        .prompt()?;

                if confirm {
                    schema.push(serde_yaml::Value::from("./.tailcallrc.graphql"));
                    let updated = serde_yaml::to_string(&yaml)?;
                    fs::write(yml_file_path, updated)?;
                }
            }
        }
    }

    Ok(())
}

fn log_endpoint_set(endpoint_set: &EndpointSet<Unchecked>) {
    let mut endpoints = endpoint_set.get_endpoints().clone();
    endpoints.sort_by(|a, b| {
        let method_a = a.get_method();
        let method_b = b.get_method();
        if method_a.eq(method_b) {
            a.get_path().as_str().cmp(b.get_path().as_str())
        } else {
            method_a.to_string().cmp(&method_b.to_string())
        }
    });
    for endpoint in endpoints {
        tracing::info!(
            "Endpoint: {} {}{} ... ok",
            endpoint.get_method(),
            API_URL_PREFIX,
            endpoint.get_path().as_str()
        );
    }
}

pub fn display_schema<W: std::io::Write>(blueprint: &Blueprint, writer: &mut W) {
    Fmt::display(Fmt::heading("GraphQL Schema:\n"), writer);
    let sdl = blueprint.to_schema();
    Fmt::display(format!("{}\n", print_schema::print_schema(sdl)), writer);
}

#[cfg(test)]
mod tests {
    use std::io::Cursor;

    use clap::Parser;

    use crate::cli::command::Cli;

    #[tokio::test]
    async fn test_run_inner() {
        let args = [
            "tailcall",
            "gen",
            "tailcall-fixtures/fixtures/generator/proto/news.proto",
            "--input",
            "proto",
            "--output",
            "graphql",
            "--resolve-ambiguous-input",
            "prefix=InPrefix_",
            "--resolve-ambiguous-output",
            "prefix=OutPrefix_",
        ];
        let cli = Cli::parse_from(args);
        let mut cursor = Cursor::new(Vec::new());
        super::run_inner(cli, &mut cursor).await.unwrap();
        let output = String::from_utf8(cursor.into_inner()).unwrap();
        insta::assert_snapshot!(output);
    }
}
