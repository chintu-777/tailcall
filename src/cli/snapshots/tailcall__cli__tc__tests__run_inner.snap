---
source: src/cli/tc.rs
expression: output
---
schema @server @upstream @link(src: "tailcall-fixtures/fixtures/generator/proto/news.proto", type: Protobuf) {
  query: Query
}

input InPrefix_NEWS_NEWS @tag(id: "news.News") {
  author: NEWS_AUTHOR
  body: String
  foo: NEWS_STATUS
  id: Int
  post_image: String
  title: String
}

input NEWS_AUTHOR @tag(id: "news.Author") {
  email: String
  name: String
}

input NEWS_MULTIPLE_NEWS_ID @tag(id: "news.MultipleNewsId") {
  ids: [NEWS_NEWS_ID]!
}

input NEWS_NEWS_ID @tag(id: "news.NewsId") {
  id: Int
}

enum NEWS_STATUS {
  DRAFT
  PUBLISHED
}

type NEWS_NEWS_LIST @tag(id: "news.NewsList") {
  news: [OutPrefix_NEWS_NEWS]!
}

type OutPrefix_NEWS_NEWS @tag(id: "news.News") {
  author: NEWS_AUTHOR
  body: String
  foo: NEWS_STATUS
  id: Int
  post_image: String
  title: String
}

type Query {
  news_add_news(news: InPrefix_NEWS_NEWS!): OutPrefix_NEWS_NEWS! @grpc(method: "news.AddNews")
  news_delete_news(news_id: NEWS_NEWS_ID!): NEWS_EMPTY! @grpc(method: "news.DeleteNews")
  news_edit_news(news: InPrefix_NEWS_NEWS!): OutPrefix_NEWS_NEWS! @grpc(method: "news.EditNews")
  news_get_all_news: NEWS_NEWS_LIST! @grpc(method: "news.GetAllNews")
  news_get_multiple_news(multiple_news_id: NEWS_MULTIPLE_NEWS_ID!): NEWS_NEWS_LIST! @grpc(method: "news.GetMultipleNews")
  news_get_news(news_id: NEWS_NEWS_ID!): OutPrefix_NEWS_NEWS! @grpc(method: "news.GetNews")
}
